### PR TITLE
add default_power_level_content_override config option.

### PR DESCRIPTION
**Pull request by andybalaam**
[Updated version of https://github.com/matrix-org/synapse/pull/12535 based on comments to the original PR.]

Add a config option that controls the default power levels in newly-created rooms.

As mentioned in the test code, the phrase "applied after" in the spec is interpreted as "if any key is present, overwrite that key completely". See https://github.com/matrix-org/matrix-spec/issues/492 for more discussion.

### CODE REVIEW

File: tests/rest/client/test_rooms.py [hunk 1]
```diff
@@ -1116,6 +1116,320 @@ def test_rooms_messages_sent(self) -> None:
...
+    def prepare(self, reactor: MemoryReactor, clock: Clock, hs: HomeServer) -> None:
+        self.admin_user_id = self.register_user("admin", "pass")
+        self.admin_access_token = self.login("admin", "pass")
+
+    def power_levels(self, room_id: str) -> Dict[str, Any]:
+        return self.helper.get_state(
+            room_id, "m.room.power_levels", self.admin_access_token
+        )
+
+    def test_default_power_levels_without_override(self) -> None:
```
Comment (anoadragon453) [inline tests/rest/client/test_rooms.py:1139]: Hmm, this test will fail every time we change the default power levels. As such, it doesn't seem very future-proof.

Likewise in the tests below. Perhaps in those tests, one should only check that the appropriate fields were added/changed?

File: tests/rest/client/test_rooms.py [hunk 2]
```diff
@@ -1116,6 +1116,320 @@ def test_rooms_messages_sent(self) -> None:
...
+                "users": {"@sid1:red": 100},
+                "users_default": 0,
+            },
+            self.power_levels(room_id),
+        )
+
+
+class RoomPowerLevelOverridesInPracticeTestCase(RoomBase):
+    """Tests that we can really do various otherwise-prohibited actions
+    based on overriding the power levels in config."""
```
Comment (anoadragon453) [inline tests/rest/client/test_rooms.py:1282]: Typically in our code style, when a docstring is multiple lines, the second `"""` will be on its own line.

(Alternatively, this used to be one line and `black` made this mistake :P)

File: tests/rest/client/test_rooms.py [hunk 3]
```diff
@@ -1116,6 +1116,320 @@ def test_rooms_messages_sent(self) -> None:
...
+    @unittest.override_config(
+        {
+            "default_power_level_content_override": {
+                "public_chat": {"events": {"custom.event": 0}},
+            }
+        },
+    )
+    def test_with_config_override_normal_user_can_post_state_event(self) -> None:
+        # Given the server has config allowing normal users to post my event type
+        # And I am a normal member of a room
```
Comment (anoadragon453) [inline tests/rest/client/test_rooms.py:1327]: nit: I'm finding the lack of punctuation between (partial) sentences here a bit odd to read.

I would normally write:

```suggestion
        # Given the server has config allowing normal users to post my event type,
        # and I am a normal member of a room
```

File: synapse/config/room.py [hunk 4]
```diff
@@ -83,4 +96,20 @@ def generate_config_section(self, **kwargs: Any) -> str:
...
+        # that they create (e.g. to send particular types of state events without
+        # needing an elevated power level).  This takes the same shape as the
+        # `power_level_content_override` parameter in the /createRoom API, but
+        # is applied before that parameter.
+        #
+        # This is something of a workaround in the absence of MSC3779 or MSC3761.
+        #
+        #default_power_level_content_override:
+        #   private_chat: { "events": { "com.example.foo" : 0 } }
+        #   trusted_private_chat: null
```
Comment (anoadragon453) [inline synapse/config/room.py:113]: And finally, could you mention that only known chat presets (`private_chat`, `trusted_private_chat` and `public_chat` are allowed here?

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Really useful feature, thank you!

File: synapse/handlers/room.py [hunk 5]
```diff
@@ -1040,9 +1040,25 @@ async def send(etype: str, content: JsonDict, **kwargs: Any) -> int:
...
 
-            # Power levels overrides are defined per chat preset
+            # If the user supplied a preset name e.g. "private_chat",
+            # we apply that preset
             power_level_content.update(config["power_level_content_override"])
 
+            # If the server config contains default_power_level_content_override,
+            # and that contains information for this room preset, apply it.
+            default_power_level_content_override = (
+                self.config.room.default_power_level_content_override
```
Comment (anoadragon453) [inline synapse/handlers/room.py:1050]: It's (unwritten) convention that for homeserver config options, we typically store them in a member variable first, and then access that member variable in the hot path. This both helps readability, as well as being perceived (though not tested) to be slightly more performant than reaching into the config dictionary.

This translates to adding a variable with a name like `self._default_power_level_content_override` to `__init__` and then using it here instead.

File: synapse/config/room.py [hunk 6]
```diff
@@ -83,4 +96,20 @@ def generate_config_section(self, **kwargs: Any) -> str:
...
+        # Override the default power levels for rooms created on this server, per
+        # room creation preset.
+        #
+        # Useful if you know that your users need special permissions in rooms
+        # that they create (e.g. to send particular types of state events without
+        # needing an elevated power level).  This takes the same shape as the
+        # `power_level_content_override` parameter in the /createRoom API, but
+        # is applied before that parameter.
+        #
+        # This is something of a workaround in the absence of MSC3779 or MSC3761.
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply synapse/config/room.py]: Done in 4309993

File: synapse/config/room.py [hunk 7]
```diff
@@ -83,4 +96,20 @@ def generate_config_section(self, **kwargs: Any) -> str:
...
+        #
+        #default_power_level_content_override:
+        #   private_chat: { "events": { "com.example.foo" : 0 } }
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply synapse/config/room.py]: Fixed in 3aed63a0e

File: tests/rest/client/test_rooms.py (hunk 2 above)
Comment (andybalaam) [reply tests/rest/client/test_rooms.py]: Fixed in 55e1eccdd

File: tests/rest/client/test_rooms.py (hunk 3 above)
Comment (andybalaam) [reply tests/rest/client/test_rooms.py:1329]: I'm vaguely using [BDD/Cucumber style](https://behave.readthedocs.io/en/stable/tutorial.html) but I'll fix it.

Comment (andybalaam) [reply tests/rest/client/test_rooms.py]: Fixed in 08b00e1b87e

File: synapse/config/room.py (hunk 4 above)
Comment (andybalaam) [reply synapse/config/room.py]: Fixed in 9adeaa7d7

File: synapse/handlers/room.py (hunk 5 above)
Comment (andybalaam) [reply synapse/handlers/room.py]: Fixed in 4dda230e4

File: tests/rest/client/test_rooms.py (hunk 1 above)
Comment (andybalaam) [reply tests/rest/client/test_rooms.py]: Fixed in ae9421caa

File: tests/rest/client/test_rooms.py (hunk 3 above)
Comment (anoadragon453) [reply tests/rest/client/test_rooms.py]: Oh, this is interesting and something I haven't come across before. Though I still think it's a little confusing given that the convention doesn't (yet) exist elsewhere in the codebase.

The lack of punctuation just confused me. Once could also write it as:

```
Given: the server has config allowing normal users to post my event type
And: I am a normal member of a room
Then: ...
```

and probably get away without it looking like odd grammar.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
LGTM otherwise!

File: tests/rest/client/test_rooms.py (hunk 3 above)
Comment (andybalaam) [reply tests/rest/client/test_rooms.py]: Are you happy with it as it is now? (I am)

File: synapse/handlers/room.py [hunk 8]
```diff
@@ -1040,9 +1040,25 @@ async def send(etype: str, content: JsonDict, **kwargs: Any) -> int:
...
+            )
+            if (
+                default_power_level_content_override
+                and default_power_level_content_override.get(preset_config)
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply synapse/handlers/room.py]: Tidied this up to be less horribly verbose, and also to check against None in 990b7fa77.

File: synapse/config/room.py [hunk 9]
```diff
@@ -100,16 +100,38 @@ def generate_config_section(self, **kwargs: Any) -> str:
...
+        #            "m.room.history_visibility": 100,
+        #            "m.room.name": 50,
+        #            "m.room.power_levels": 100,
+        #            "m.room.server_acl": 100,
+        #            "m.room.tombstone": 100
+        #        },
+        #        "events_default": 1
+        #    }
+        #    trusted_private_chat: null
+        #    public_chat: null
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply synapse/config/room.py]: Fixed in a91a04326ac63

File: changelog.d/12618.feature [hunk 10]
```diff
@@ -0,0 +1 @@
+Add a `default_power_level_content_override` config option to set default room power levels.
```
Comment (anoadragon453) [inline changelog.d/12618.feature:1]:
```suggestion
Add a `default_power_level_content_override` config option to set default room power levels per room preset.
```

File: synapse/config/room.py [hunk 11]
```diff
@@ -83,4 +96,40 @@ def generate_config_section(self, **kwargs: Any) -> str:
...
+        # Valid keys are some or all of `private_chat`, `trusted_private_chat`
+        # and `public_chat`. Inside each of those should be any of the
+        # properties allowed in `power_level_content_override` in the
+        # /createRoom API. If any property is missing, its default value will
+        # continue to be used. If any property is present, it will overwrite
+        # the existing default completely (so if the `events` property exists,
+        # the default event power levels will be ignored).
+        #
+        #default_power_level_content_override:
+        #    private_chat: {
```
Comment (anoadragon453) [inline synapse/config/room.py:121]:
```suggestion
        #    "private_chat": {
```

for consistency.

File: synapse/config/room.py [hunk 12]
```diff
@@ -83,4 +96,40 @@ def generate_config_section(self, **kwargs: Any) -> str:
...
+        #            "m.room.canonical_alias": 50,
+        #            "m.room.encryption": 100,
+        #            "m.room.history_visibility": 100,
+        #            "m.room.name": 50,
+        #            "m.room.power_levels": 100,
+        #            "m.room.server_acl": 100,
+        #            "m.room.tombstone": 100
+        #        },
+        #        "events_default": 1
+        #    }
```
Comment (anoadragon453) [inline synapse/config/room.py:134]: Sorry, quick thing I've only noticed just now. This isn't valid YAML. Using `{` / `}` syntax will only work in YAML if it not split across multiple lines.

The below should work though (and is more YAML-y):

```suggestion
        #default_power_level_content_override:
        #    private_chat:
        #        "events":
        #            "com.example.myeventtype" : 0
        #            "m.room.avatar": 50
        #            "m.room.canonical_alias": 50
        #            "m.room.encryption": 100
        #            "m.room.history_visibility": 100
        #            "m.room.name": 50
        #            "m.room.power_levels": 100
        #            "m.room.server_acl": 100
        #            "m.room.tombstone": 100
        #        "events_default": 1
```

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Apologies, minor items that came up during a final review.

File: synapse/config/room.py (hunk 12 above)
Comment (andybalaam) [reply synapse/config/room.py]: Fixed in 8a639df70. Wow, I thought I already hated YAML.

File: docs/sample_config.yaml [hunk 13]
```diff
@@ -2476,20 +2476,18 @@ push:
...
-#        "events": {
-#            "com.example.myeventtype" : 0,
-#            "m.room.avatar": 50,
-#            "m.room.canonical_alias": 50,
-#            "m.room.encryption": 100,
-#            "m.room.history_visibility": 100,
-#            "m.room.name": 50,
-#            "m.room.power_levels": 100,
-#            "m.room.server_acl": 100,
+#    private_chat:
```
Comment (anoadragon453) [inline docs/sample_config.yaml:2479]:
```suggestion
#    "private_chat":
```

File: synapse/config/room.py [hunk 14]
```diff
@@ -118,18 +118,16 @@ def generate_config_section(self, **kwargs: Any) -> str:
...
-        #        "events": {
-        #            "com.example.myeventtype" : 0,
-        #            "m.room.avatar": 50,
-        #            "m.room.canonical_alias": 50,
-        #            "m.room.encryption": 100,
-        #            "m.room.history_visibility": 100,
-        #            "m.room.name": 50,
-        #            "m.room.power_levels": 100,
-        #            "m.room.server_acl": 100,
+        #    private_chat:
```
Comment (anoadragon453) [inline synapse/config/room.py:121]:
```suggestion
        #    "private_chat":
```

**Review by anoadragon453 — APPROVED** [review]
LGTM assuming CI is happy!

**Pull request merged**